doc/source/cluster/kubernetes/examples/distributed-checkpointing-with-gcsfuse.md

```diff
@@ -0,0 +1,297 @@
+BUCKET=<your GCS BUCKET>
+gcloud storage buckets add-iam-policy-binding gs://${BUCKET} --member "principal://iam.googleapis.com/projects/${PROJECT_NUMBER}/locations/global/workloadIdentityPools/${PROJECT_ID}.svc.id.goog/subject/ns/default/sa/pytorch-distributed-training"  --role "roles/storage.objectUser"
+```
+
+See [Access Cloud Storage buckets with the Cloud Storage FUSE CSI driver](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver) for more details.
+
+## Deploy the RayJob
+
+Download the RayJob that executes all the steps documented in [Finetuning a Pytorch Image Classifier with Ray Train](https://docs.ray.io/en/latest/train/examples/pytorch/pytorch_resnet_finetune.html). The [source code](https://github.com/ray-project/kuberay/tree/master/ray-operator/config/samples/pytorch-resnet-image-classifier) is also in the KubeRay repository.
+
+```
+curl -LO https://raw.githubusercontent.com/ray-project/kuberay/master/ray-operator/config/samples/pytorch-resnet-image-classifier/ray-job.pytorch-image-classifier.yaml
+```
+
+Modify the RayJob by replacing all instances of the `GCS_BUCKET` placeholder with the Google Cloud Storage bucket you created earlier. Alternatively you can use `sed`:
+```
+sed -i "s/GCS_BUCKET/<your GCS bucket>/g" ray-job.pytorch-image-classifier.yaml
+```
+
+Deploy the RayJob
+```
+kubectl create -f ray-job.pytorch-image-classifier.yaml
+```
+
+The deployed RayJob includes the following configuration to enable distributed checkpointing to a shared filesystem:
+* 4 Ray workers, each with a single GPU.
+* All Ray nodes use the `pytorch-distributed-training` ServiceAccount which we created earlier.
+* Includes volumes that are managed by the `gcsfuse.csi.storage.gke.io` CSI driver.
+* Mounts a shared storage path `/mnt/cluster_storage`, backed by the GCS bucket you created earlier.
+
+The Pod can be configured with annotations which allows for finer grain control of the GCSFuse sidecar container. See [Specify Pod annotations](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#pod-annotations) for more details.
+```
+annotations:
+  gke-gcsfuse/volumes: "true"
+  gke-gcsfuse/cpu-limit: "0"
+  gke-gcsfuse/memory-limit: 5Gi
+  gke-gcsfuse/ephemeral-storage-limit: 10Gi
+```
+
+Mount options are also specified when defining the GCSFuse container volume:
+```
+csi:
+  driver: gcsfuse.csi.storage.gke.io
+  volumeAttributes:
+    bucketName: GCS_BUCKET
+    mountOptions: "implicit-dirs,uid=1000,gid=100"
+```
+
+See [Mount options](https://cloud.google.com/kubernetes-engine/docs/how-to/persistent-volumes/cloud-storage-fuse-csi-driver#mount-options) to learn more about mount options.
+
+Logs from the Ray job should indicate use of the shared remote filesystem in `/mnt/cluster_storage` and the checkpointing directory. For example:
+```
+Training finished iteration 10 at 2024-04-29 10:22:08. Total running time: 1min 30s
```

Comment:
The job fails; I will troubleshoot the issue.

<img width="1440" alt="Screenshot 2024-05-17 at 1 18 38 PM" src="https://github.com/ray-project/ray/assets/20109646/749a0397-28ef-45ce-99c6-783cc3babf15">

Reply:
I didn’t have a chance to dive deep into troubleshooting this issue, but it seems to be a GCSFuse bug (or a lack of observability). I tried to create a file in `/mnt/cluster_storage`, and although the command didn't report an error, I couldn’t see the file.

![Screenshot 2024-05-17 at 2 00 29 PM](https://github.com/ray-project/ray/assets/20109646/a071cef8-b07b-49c5-bf19-1aa9ef374bae)

Reply:
Update: Synced with @andrewsykim offline. We need to set the GCS bucket's permissions to 'Uniform' instead of fine-grained.